### PR TITLE
Made improvements to pretty printing

### PR DESCRIPTION
**Pull request by BluntSporks**
Hi, I found this PHP-Parser project and it appears quite useful. I started using the PHP Pretty Printer, and wrote a number of changes to this feature that I think are improvements. They are in this changeset that I am making available to you in case you want to pull the changes (or any part of them that you find useful).

Summary of the changes:
1. Wrote a fixTokens() function to fix the PHP Tokens so that any space found attached to a PHP end tag, ?>, is instead moved to the HTML following the end tag so it is preserved.
2. Saved information about heredoc and nowdoc tags in the parser so that the actual heredoc and nowdoc tags can be preserved on request.
3. Added a name callback function to transform constant names, for example, all instances of TRUE, FALSE, and NULL can be transformed to true, false, and null by using strtolower() as the callback.
4. Allowing strings that end in newlines to encode the newline because it looks better, for example, "\n" or "This is an example\n" will keep the \n sequence.
5. Rewrote the array printing function so that the arrays are nicely wrapped and easier to read.
6. Rewrote how the open brace and close brace are handled, so that the formatting of open brace and close brace can be changed systematically just by setting some configuration parameters. This is useful when following different coding styles that mandate different types of indentation and brace syntax.
7. Allowed blank lines before comments if a configuration parameter is set.
8. Changed the prettyPrintFile() to optionally include the closing tag if that is requested, since some standards require that.
9. Wrote a wrapLines() function that takes prettified code and wraps it after specified token types to within a given line length, since most formatting standards require that. I thought a separate function was easier to understand than modifying all of the existing functions.

I can give you some examples of before and after code and show you how much this is improved in terms of flexibility and appearance of output.

The rest of this message consists of a pretty printer script that I wrote which uses most of these new features.

``` php
#!/usr/bin/php
<?php

/**
 * This is a pretty-printer using PHP Parser
 */
if ($argc != 2) {
   die("Usage: {$argv['0']} [filename]\n");
}
$fileName = $argv[1];
if (!file_exists($fileName)) {
   die("File {$fileName} not found\n");
}
require '/home/dsg/library/PHP-Parser/lib/bootstrap.php';
ini_set('xdebug.max_nesting_level', 2000);
$sourceCode = file_get_contents($fileName);
$parser = new PhpParser\Parser(new PhpParser\Lexer());
$prettyPrinter = new PhpParser\PrettyPrinter\Standard();

// Configure general formatting
$prettyPrinter->setIndentString('   ');
$prettyPrinter->setUseBlankBeforeComment(true);
$prettyPrinter->setUseArrayIndent(true);
$prettyPrinter->setPreserveHeredoc(true);
$prettyPrinter->setNameCallback('strtolower');

// Configure string length and line wrap
$prettyPrinter->setLineWrap(120);
$prettyPrinter->addWrapToken(',');
$prettyPrinter->addWrapToken('.');
$prettyPrinter->addWrapToken(T_BOOLEAN_AND);
$prettyPrinter->addWrapToken(T_BOOLEAN_OR);
$prettyPrinter->addWrapToken(T_LOGICAL_AND);
$prettyPrinter->addWrapToken(T_LOGICAL_OR);
$prettyPrinter->addWrapToken(T_LOGICAL_XOR);

// Configure line breaks before keywords
$prettyPrinter->setKeywordNewline('catch');
$prettyPrinter->setKeywordNewline('else');
$prettyPrinter->setKeywordNewline('elseif');
$prettyPrinter->setKeywordNewline('finally');
$prettyPrinter->setKeywordNewline('while');
try {

   // Parse
   $stmts = $parser->parse($sourceCode);

   // Pretty print
   $prettyCode = $prettyPrinter->prettyPrintFile($stmts, true);
   $wrappedCode = $prettyPrinter->wrapLines($prettyCode);
   file_put_contents($fileName, $wrappedCode);
   echo "Wrote {$fileName}\n";
}
catch (PhpParser\Error $e) {
   echo "Error parsing {$fileName}: ", $e->getMessage();
}
?>
```

### CODE REVIEW

**Comment by GrahamCampbell** [comment]
Any news on this?

File: grammar/kmyacc.php.parser [hunk 1]
```diff
@@ -11,7 +11,7 @@ namespace PhpParser;
 /* This is an automatically GENERATED file, which should not be manually edited.
  * Instead edit one of the following:
  *  * the grammar file grammar/zend_language_parser.phpy
```
Comment (hacfi) [inline grammar/kmyacc.php.parser]: You could remove the `y` from `phpy` if you commit again

**Comment by ovr** [comment]
Any news on this?

**Comment by BluntSporks** [comment]
I thought they weren't interested in the updates and already deleted my
repo. So don't worry about it.

**Comment by joelwurtz** [comment]
For me this should not be the role of this library to pretty print something as this will imply at lot of complexity here.

I have resolved this by piping the generated code through https://github.com/FriendsOfPHP/PHP-CS-Fixer and it works well.

**Comment by TomasVotruba** [comment]
@joelwurtz That's the same conclusion I came in. Do you still use it?

Now you can use `FormatPerservingPrinter`: https://github.com/nikic/PHP-Parser/issues/41#issuecomment-269230733

I use it and it works well apart few glitches: https://github.com/nikic/PHP-Parser/issues/400

**Comment by m1guelpf** [comment]
@nikic I believe this can be closed now...

**Comment by nikic** [comment]
Right, realistically this is not landing at this point. My current stance on the issue is that for non-trivial adjustment to the formatting an existing formatting library should be used, as mentioned in https://github.com/nikic/PHP-Parser/pull/116#issuecomment-238037924.

**Pull request closed without merging**